### PR TITLE
Make dspy.Evaluate Typed and DSPy Only (No dsp)

### PR DESCRIPTION
**Pull request by KCaverly**
In an effort to harden the Evaluate function, I have moved all metrics from dsp to dspy, added type hints and made Evaluate Pydantic validated. This should not affect any fundamental functionality, and is simply an effort to increase code quality and coverage.

### CODE REVIEW

File: dspy/evaluate/evaluate.py [hunk 1]
```diff
@@ -20,50 +25,52 @@ def HTML(x):
...
 """
 TODO: Counting failures and having a max_failure count. When that is exceeded (also just at the end),
 we print the number of failures, the first N examples that failed, and the first N exceptions raised.
 """
 
+Metric = t.Callable[[Example, Prediction, ...], bool]
+
+
+class Evaluate(BaseModel):
+    # Example is not yet Pydantic valid
```
Comment (arnavsinghvi11) [inline dspy/evaluate/evaluate.py:37]: can remove this comment now :)

File: dspy/evaluate/evaluate.py [hunk 2]
```diff
@@ -20,50 +25,52 @@ def HTML(x):
...
+        display_progress: bool,
+    ) -> t.Tuple[list[Example], int, int]:
         ncorrect = 0
         ntotal = 0
         reordered_devset = []
 
         pbar = tqdm.tqdm(total=len(devset), dynamic_ncols=True, disable=not display_progress, file=sys.stdout)
         for idx, arg in devset:
             with logging_redirect_tqdm():
+                print(f"IDX: {idx}, {arg}, {type(idx)}, {type(arg)}")
```
Comment (arnavsinghvi11) [inline dspy/evaluate/evaluate.py:73]: should this be printed or logged and outputted in a verbose/debug setting?

File: dspy/evaluate/metrics.py [hunk 3]
```diff
@@ -1,22 +1,76 @@
...
+
+def passage_has_answers(passage: str, answers: list[str]) -> bool:
+    return has_answer(
+        tokenized_answers=[DPR_normalize(normalize_text(ans)) for ans in answers],
+        text=normalize_text(passage),
+    )
+
+
+def answer_exact_match(example: Example, pred: Prediction, frac: float = 0.90, *_args, **_kwargs) -> bool:
+    if not isinstance(example.answer, (str, list)):
```
Comment (arnavsinghvi11) [inline dspy/evaluate/metrics.py:60]: fine to leave as original [assert statement](https://github.com/stanfordnlp/dspy/blob/d7dace6b51e1c0899fd798c575262310939f6deb/dspy/evaluate/metrics.py#L7)?

File: dspy/evaluate/metrics.py [hunk 4]
```diff
@@ -1,22 +1,76 @@
...
+        raise ValueError("example.answer must be str or list")
+
+    if isinstance(example.answer, str):
+        return answer_match(pred.answer, [example.answer], frac=frac)
+
+    return answer_match(pred.answer, example.answer, frac=frac)
+
+
+def answer_passage_match(example: Example, pred: Prediction, *_args, **_kwargs):
+    if not isinstance(example.answer, (str, list)):
```
Comment (arnavsinghvi11) [inline dspy/evaluate/metrics.py:70]: same comment as above

File: dspy/evaluate/metrics.py [hunk 5]
```diff
@@ -1,22 +1,76 @@
...
 
 
-def answer_exact_match(example, pred, trace=None, frac=1.0):
-    assert(type(example.answer) is str or type(example.answer) is list)
-    
-    if type(example.answer) is str:
-        return dsp.answer_match(pred.answer, [example.answer], frac=frac)
-    else: # type(example.answer) is list
-        return dsp.answer_match(pred.answer, example.answer, frac=frac)
+def em(prediction: str, answers_list: list[str]) -> bool:
```
Comment (arnavsinghvi11) [inline dspy/evaluate/metrics.py:10]: missing [asserts](https://github.com/stanfordnlp/dspy/blob/d7dace6b51e1c0899fd798c575262310939f6deb/dsp/utils/metrics.py#L10)?

File: dspy/evaluate/metrics.py [hunk 6]
```diff
@@ -1,22 +1,76 @@
...
-answer_exact_match_str = dsp.answer_match
 
-def answer_passage_match(example, pred, trace=None):
-    assert(type(example.answer) is str or type(example.answer) is list)
-    
-    if type(example.answer) is str:
-        return dsp.passage_match(pred.context, [example.answer])
-    else: # type(example.answer) is list
-        return dsp.passage_match(pred.context, example.answer)
+def f1(prediction: str, answers_list: list[str]) -> float:
```
Comment (arnavsinghvi11) [inline dspy/evaluate/metrics.py:14]: [asserts](https://github.com/stanfordnlp/dspy/blob/d7dace6b51e1c0899fd798c575262310939f6deb/dsp/utils/metrics.py#L16)

**Comment by arnavsinghvi11** [comment]
Thanks for this PR @KCaverly ! As we're moving the metrics from dsp to dspy, it seems like we are missing some metrics in [there](https://github.com/stanfordnlp/dspy/blob/main/dsp/utils/metrics.py) (novel_f1_score, precision_score, hotpot_f1_score, HotPotF1, nF1). Granted, these are not used in the repo but maybe we can keep them for now until a greater refactor (similar to the case with datasets). 

We can probably revisit this later, but with this change, should be good to remove them from the dsp/ folder now.

Also feel free to check PR #935 since there may be some overlap/changes.

**Comment by thomasahle** [comment]
LGTM

**Pull request closed without merging**